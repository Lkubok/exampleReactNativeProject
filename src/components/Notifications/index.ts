import Component from './Notifications';

export default Component;
