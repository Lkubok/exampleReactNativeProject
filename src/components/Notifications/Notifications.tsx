/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable no-console */
/* eslint-disable @typescript-eslint/require-await */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable react-native/no-color-literals */
/* eslint-disable react-native/no-inline-styles */
import React, { useEffect, useState, useRef, useMemo, useCallback } from 'react';
import { Platform } from 'react-native';
import Bugsnag from '@bugsnag/react-native';
import { Notification } from 'react-native-in-app-message';
import _ from 'lodash';
import messaging from '@react-native-firebase/messaging';
import { useSelector } from 'react-redux';

import { NotificationComponent } from 'components';
import { usePushNotification } from 'hooks';
import { variables } from 'res';
import { opacityColor } from 'utils';
import { selectors as communicationSelectors } from 'store/communication';

const handleFcmRequestFailure = async (error: any) => {
  try {
    if (error !== messaging.AuthorizationStatus.DENIED && _.isError(error)) {
      Bugsnag.notify(error, (report) => {
        // eslint-disable-next-line no-param-reassign
        report.context = 'Problem with FCM request.';
      });
    }
  } catch (e) {
    return null;
  }
};
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const Notifications = () => {
  const ref = useRef<any>(null);
  const timerRef = useRef<any>(null);
  const [content, setContent] = useState<any>({ body: ' ', title: ' ' });
  const { setClickedNotificationType, setNotificationData, setTriggerMessageThreadUpdate } = usePushNotification();
  const currentThreadParams = useSelector(communicationSelectors.getCurrentThreadParams);

  useEffect(() => {
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    messaging()
      .hasPermission()
      .then((authStatus) => {
        const enabled =
          authStatus === messaging.AuthorizationStatus.AUTHORIZED ||
          authStatus === messaging.AuthorizationStatus.PROVISIONAL;
        if (!enabled) {
          messaging().requestPermission().catch(handleFcmRequestFailure);
        }
      });

    messaging()
      .getToken()
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      .then((to) => {
        console.log('FCM TOKEN', to);
      })
      .catch((e) => console.error('ERR', e));

    messaging()
      .getInitialNotification()
      .then(handleNotificationClick)
      .catch(() => {});
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    const unsubscribe = messaging().onNotificationOpenedApp(handleNotificationClick);
    return unsubscribe;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const handleNotificationClick = useCallback(
    (payload: any) => {
      let data;
      const notification = payload.notification || payload;

      if (Platform.OS === 'android') {
        data = payload?.data;
      } else {
        data = payload?.data;
      }

      const unifiedNotificationPayload = { ...notification, ...data };
      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
      setNotificationData(unifiedNotificationPayload);
      setClickedNotificationType(unifiedNotificationPayload.notification_type);

      return null;
    },
    [setClickedNotificationType, setNotificationData],
  );

  useEffect(() => {
    // eslint-disable-next-line @typescript-eslint/require-await
    const unsubscribe = messaging().onMessage(async (remoteMessage) => {
      const shouldPushNotificaionRefreshMessagingThread =
        currentThreadParams?.customerSlug === remoteMessage?.data?.customer_slug &&
        currentThreadParams?.offerSlug === remoteMessage?.data?.offer_slug &&
        remoteMessage.data?.notification_type === 'new_message';

      if (shouldPushNotificaionRefreshMessagingThread) {
        setContent(remoteMessage);
        setTriggerMessageThreadUpdate(true);
      } else {
        ref.current?.hide();
        setContent(remoteMessage);
        ref.current?.show();
        timerRef.current = setTimeout(() => {
          ref.current?.hide();
        }, 5000);
      }
    });
    return unsubscribe;
  }, [currentThreadParams?.customerSlug, currentThreadParams?.offerSlug, setTriggerMessageThreadUpdate]);

  const onPress = useCallback(() => {
    ref.current?.hide();
    handleNotificationClick(content);
  }, [content, handleNotificationClick]);

  const customComponent = useMemo(
    () => (
      <NotificationComponent
        title={_.get(content, 'notification.title', content.title)}
        body={_.get(content, 'notification.body', content.body)}
        onPress={onPress}
      />
    ),
    [content, onPress],
  );

  const topInset = Platform.OS === 'android' ? 32 : 0;
  const backgroundColor = Platform.OS === 'android' ? opacityColor(variables.colors.white, 0.8) : 'transparent';

  return (
    <Notification
      autohide={false}
      tapticFeedback={true}
      blurType="light"
      customComponent={customComponent}
      hideStatusBar={false}
      style={{ backgroundColor, borderRadius: 6, marginTop: topInset }}
      ref={ref}
      showKnob
    />
  );
};

export default Notifications;
