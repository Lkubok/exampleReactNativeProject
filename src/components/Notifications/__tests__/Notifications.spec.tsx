import React from 'react';
import { act, customRender } from 'utils/testUtils';

import Notifications from '../Notifications';

jest.useFakeTimers();

describe('Notifications component', () => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let props: any;
  beforeEach(() => {
    props = {
      disabled: true,
      loading: false,
      onPress: () => {},
      title: 'testingNotifications',
      useNativeDriver: true,
    };
  });

  it('should match snapshot', () => {
    const snapshot = customRender(<Notifications {...props} />).toJSON();
    // eslint-disable-next-line no-void
    void act(() => {
      expect(snapshot).toMatchSnapshot();
    });
  });
});
