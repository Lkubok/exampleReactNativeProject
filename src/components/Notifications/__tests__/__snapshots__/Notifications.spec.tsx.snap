// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications component should match snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={false}
    forwardedRef={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onLayout={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderRadius": 14,
        "elevation": 2,
        "justifyContent": "center",
        "position": "absolute",
        "shadowColor": "[7m[38;2;0;0;0m#000[39m[27m",
        "shadowOffset": Object {
          "height": 5,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 4,
        "top": 22,
        "transform": Array [
          Object {
            "translateY": -9000,
          },
        ],
        "width": "96.5%",
        "zIndex": 2,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
          "borderRadius": 6,
          "marginTop": 0,
          "width": "100%",
        }
      }
    >
      <View
        collapsable={false}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
      >
        <Blur
          blurAmount={7}
          blurType="light"
          pointerEvents="none"
          style={
            Object {
              "borderRadius": 6,
              "height": "100%",
              "position": "absolute",
              "width": "100%",
            }
          }
        />
        <View
          enabled={false}
          minForce={0.2}
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "height": "100%",
                "justifyContent": "center",
                "width": "100%",
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              hitSlop={
                Object {
                  "bottom": 12,
                  "left": 12,
                  "right": 12,
                  "top": 12,
                }
              }
              onClick={[Function]}
              style={
                Object {
                  "flexDirection": "row",
                  "height": 100,
                  "opacity": 1,
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "marginHorizontal": 12,
                  }
                }
              >
                <Image
                  resizeMode="contain"
                  source={
                    Object {
                      "testUri": "../../../src/assets/notificationIcon.png",
                    }
                  }
                  style={
                    Object {
                      "borderRadius": 30,
                      "height": 40,
                      "width": 40,
                    }
                  }
                />
              </View>
              <View
                style={
                  Object {
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "[7m[38;2;0;0;0m#000000[39m[27m",
                      "fontFamily": "SFCompactText-Bold",
                      "fontSize": 18,
                      "marginBottom": 12,
                    }
                  }
                >
                   
                </Text>
                <Text
                  style={
                    Object {
                      "color": "[7m[38;2;0;0;0m#000000[39m[27m",
                      "fontFamily": "SFCompactText-Regular",
                      "fontSize": 17,
                    }
                  }
                >
                   
                </Text>
              </View>
            </View>
            <View
              style={
                Object {
                  "backgroundColor": "[7m[38;2;0;0;0m#000[39m[27m",
                  "borderRadius": 4,
                  "height": 4,
                  "marginVertical": 6,
                  "opacity": 0.3,
                  "width": 50,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
